middle_end/flambda/simplify/env/rec_uses.ml

```diff
@@ -0,0 +1,375 @@
+    let map = Continuation.Map.add cont elt t.map in
+    { t with stack; map; }
+
+let update_top_of_stack ~t ~f =
+  match t.stack with
+  | [] -> Misc.fatal_errorf "empty stack of variable uses in flambda2"
+  | elt :: stack -> { t with stack = f elt :: stack; }
+
+let add_binding var name_occurrences t =
+  update_top_of_stack ~t ~f:(fun elt ->
+    let bindings =
+      Variable.Map.update var (function
+        | None -> Some name_occurrences
+        | Some _ ->
+          Misc.fatal_errorf "The same variable has been bound twice"
+      ) elt.bindings
+    in
+    let generate_phantom_lets =
+      !Clflags.debug && !Clflags.Flambda.Expert.phantom_lets
+    in
+    let used_in_handler =
+      if Variable.user_visible var && generate_phantom_lets then
+        Name_occurrences.add_variable elt.used_in_handler var Name_mode.phantom
+      else
+        elt.used_in_handler
+    in
+    { elt with bindings; used_in_handler; }
+  )
+
+let add_used_in_current_handler name_occurrences t =
+  update_top_of_stack ~t ~f:(fun elt ->
+    let used_in_handler =
+      Name_occurrences.union elt.used_in_handler name_occurrences
+    in
+    { elt with used_in_handler; }
+  )
+
+let add_apply_result_cont k t =
+  update_top_of_stack ~t ~f:(fun elt ->
+    let apply_result_conts = Continuation.Set.add k elt.apply_result_conts in
+    { elt with apply_result_conts; }
+  )
+
+let add_apply_cont_args cont arg_name_occurrences t =
+  update_top_of_stack ~t ~f:(fun elt ->
+    let apply_cont_args =
+      Continuation.Map.update cont (fun map_opt ->
+        let map = Option.value ~default:Numbers.Int.Map.empty map_opt in
+        let map, _ = List.fold_left (fun (map, i) name_occurrences ->
+          let map =
+            Numbers.Int.Map.update i (fun old_opt ->
+              let old = Option.value ~default:Name_occurrences.empty old_opt in
+              Some (Name_occurrences.union old name_occurrences)
+            ) map
+          in
+          map, i + 1
+          ) (map, 0) arg_name_occurrences
+        in
+        Some map
+      ) elt.apply_cont_args
+    in
+    { elt with apply_cont_args; }
+  )
+
+(* Variable graph *)
+(* ************** *)
+
+module Var_graph = struct
+
+  type t = Variable.Set.t Variable.Map.t
+
+  let print ppf t =
+    Variable.Map.print Variable.Set.print ppf t
+
+  let empty : t = Variable.Map.empty
+
+  let add_edge ~src ~dst t =
+    Variable.Map.update src (function
+      | None -> Some (Variable.Set.singleton dst)
+      | Some set -> Some (Variable.Set.add dst set)
+    ) t
+
+  let edges ~src t =
+    match Variable.Map.find src t with
+    | res -> res
+    | exception Not_found -> Variable.Set.empty
+
+  (* breadth-first reachability analysis
+     CR gbury: would using sets (and thus union and diff) instead of a
+               queue be better ? Enqueuing new nodes would be faster, but
+               we'd lose the bread-first aspect, and revert to some kind of
```

Comment:
Always go for the bread first !

Reply:
I've switched to using an `enqueued` set rather than the `visited` one, and removed the comment, ^^